### PR TITLE
Add optional actions (buttons) to notifications

### LINKED ISSUE #44: Add optional action buttons to notifications
[Android 4.1+][1], [iOS 8+][2], [Chrome Apps][3] and [Mozilla XUL][4] all support showing action buttons at the bottom of a notification. This allows users to quickly handle the most common tasks for a particular notification, without having to open the originating application.

[1]: http://developer.android.com/design/patterns/notifications.html#actions
[2]: https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIUserNotificationAction_class/
[3]: https://developer.chrome.com/apps/notifications#property-NotificationOptions-buttons
[4]: https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XUL/Method/appendNotification

We'd like to add action buttons to the Notifications API, and implement them in Chrome.

Strawman API:
```webidl
dictionary NotificationAction {
    DOMString id;  // If clicked, this will become NotificationEvent.actionId
    DOMString title;  // Button text
}
partial dictionary NotificationOptions {
    sequence<NotificationAction> actions;
}
partial interface Notification {
    FrozenArray<NotificationAction> actions;
}
partial interface NotificationEvent {
    DOMString actionId;  // Id of clicked action, or null if user clicked notification body instead
}
```

Action support would be feature detected via `Notification.prototype.hasOwnProperty('actions')`. UAs which expose that must support up to 2 actions. If more than 2 actions are provided in `NotificationOptions.actions`, throw a TypeError. The number 2 was chosen because it's the maximum supported by iOS and Chrome OS, and although Android supports 3 buttons, Chrome for Android already uses one of the available slots to show a link to per-origin settings. (If later UAs want to support more actions, we could add a `maxSupportedActions` constant to Notification).

Presentation of actions would be implementation-defined, for example on Android action buttons are only shown when the notification is [expanded](http://developer.android.com/design/patterns/notifications.html#ExpandedLayouts). Hence websites should not rely on actions, and should provide the same functionality via alternate UI that can be reached by clicking on the notification body.

I explicitly left out action icons, since only Android supports them, and Android currently only supports icon resources shipped in a native app's apk (not icons downloaded at runtime), so they wouldn't be implementable yet there either.

@sicking @beverloo

### PR DESCRIPTION
**Pull request by johnmellor**
Resolves #44

### CODE REVIEW

**Comment by johnmellor** [comment]
[HTML diff](http://services.w3.org/htmldiff?doc1=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2F90e29647469fb26b1f707cf93d8c787b92d2bbdc%2Fnotifications.html&doc2=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2Fa74688eb593d6040173a77907879db47b96ce3c1%2Fnotifications.html) to aid reviewing.

File: notifications.bs [hunk 1]
```diff
@@ -83,6 +83,14 @@ associated <dfn>icon URL</dfn>, <dfn>icon resource</dfn>, <dfn>sound URL</dfn>,
 <p class=note>Developers are encouraged to not convey information through an icon, sound,
 or vibration pattern that is not otherwise accessible to the end user.
 
+<p>A <a lt="concept notification">notification</a> has an associated list of
+<dfn>actions</dfn> (which may be empty). Platforms may render these as buttons.
```
Comment (beverloo) [inline notifications.bs]: nit: You could remove the parenthesis by rephrasing as "... zero or more actions."

File: notifications.bs [hunk 2]
```diff
@@ -83,6 +83,14 @@ associated <dfn>icon URL</dfn>, <dfn>icon resource</dfn>, <dfn>sound URL</dfn>,
...
+Users may activate these, as alternatives to activating the notification itself.
+
+<p class=note>Since action display is platform-dependent, developers are
```
Comment (annevk) [inline notifications.bs]: "display of actions"

File: notifications.bs (hunk 1 above)
Comment (beverloo) [inline notifications.bs]: "Platforms may render these as buttons."

Let's stick to normative text in this paragraph, and move the display suggestions to the note below. This sentence is more like an example anyway.

Comment (annevk) [inline notifications.bs]: I prefer "action list" as single `<dfn>`. And then "consisting of zero or more actions".

File: notifications.bs [hunk 3]
```diff
@@ -227,16 +241,22 @@ section of HTML. [[!HTML]]
...
-U+000A LINE FEED (LF) characters. For each paragraph of the
-<a lt="concept title">title</a> and <a>body</a>, the
-<a lt="concept notification">notification</a>'s
+<a lt="concept title">title</a>, <a>body</a> and each of the <a>actions</a>.
+Each is expected to be treated as an independent set of one or more
+bidirectional algorithm paragraphs when displayed, as defined by the
+bidirectional algorithm's rules P1, P2, and P3, including, for instance,
+supporting the paragraph-breaking behavior of U+000A LINE FEED (LF) characters.
+For each paragraph of the <a lt="concept title">title</a>, <a>body</a> and each
+of the <a>actions</a>, the <a lt="concept notification">notification</a>'s
```
Comment (beverloo) [inline notifications.bs]: Just the action's "title" attribute, BiDi doesn't have to care about the "action" value.

File: notifications.bs [hunk 4]
```diff
@@ -426,6 +456,8 @@ interface Notification : EventTarget {
   static readonly attribute NotificationPermission permission;
   [Exposed=Window] static void requestPermission(optional NotificationPermissionCallback callback);
 
+  static readonly attribute unsigned long maxActions;
```
Comment (annevk) [inline notifications.bs]: We generally don't expose limits. See e.g. vibration pattern.

File: notifications.bs [hunk 5]
```diff
@@ -728,10 +800,12 @@ partial interface ServiceWorkerRegistration {
  Exposed=ServiceWorker]
 interface NotificationEvent : ExtendableEvent {
   readonly attribute Notification notification;
+  readonly attribute DOMString? action;
 };
 
 dictionary NotificationEventInit : ExtendableEventInit {
   required Notification notification;
+  DOMString? action = null;
```
Comment (annevk) [inline notifications.bs]: I would prefer special casing the empty string. That would match `tag` better iirc.

**Comment by annevk** [comment]
Thank you for this PR. I left some comments.

File: notifications.bs [hunk 6]
```diff
@@ -246,9 +266,10 @@ of rules P2 and P3 if it has a value other than "<code>auto</code>". [[!BIDI]]
 <p>The <a lt="concept notification">notification</a>'s
 <a lt="concept language">language</a> specifies the primary language for the
 <a lt="concept notification">notification</a>'s
-<a lt="concept title">title</a> and <a>body</a>. Its value is a string. The empty string
-indicates that the primary language is unknown. Any other string must be interpreted as a
-language tag. Validity or well-formedness are not enforced. [[!LANG]]
+<a lt="concept title">title</a>, <a>body</a> and <a>actions</a>. Its value is a
```
Comment (beverloo) [inline notifications.bs]: Again: Just the titles of the actions.

File: notifications.bs (hunk 4 above)
Comment (beverloo) [inline notifications.bs]: @sicking also suggested this in #44. Availability of actions might influence how a page emphasizes alternatives to actions on the page displayed in response to activating the notification.

That said, I'm having mixed feelings about only storing `maxActions` actions in the action list.

File: notifications.bs [hunk 7]
```diff
@@ -647,20 +690,49 @@ getter must return the <a lt="concept notification">notification</a>'s
...
+objects dispatch events during their lifecycle, which developers can use to
+generate desired behaviors.
 
 <p>The <code>click</code> event dispatches when the user activates a
 notification.
 
 <pre class=example>
-var not = new Notification("Gebrünn Gebrünn by Paul Kalkbrenner", { icon: "newsong.svg", tag: "song" })
-notification.onclick = function() { displaySong(this) }</pre>
+var not = new Notification("Gebrünn Gebrünn by Paul Kalkbrenner", { icon: "newsong.svg", tag: "song" });
```
Comment (beverloo) [inline notifications.bs]: s/.svg/.png/. At least in Chrome we don't support SVG images, and I imagine other implementations will also have difficulties with this.

Comment (annevk) [inline notifications.bs]: You should though, so I don't support changing that (not as part of this PR anyway).

Comment (beverloo) [inline notifications.bs]: Oh, I see that John's not actually changing the line. Let's leave that for a different issue :).

(The spec doesn't say anything about supported image formats. Supporting SVG will be rather complicated for us.)

**Comment by annevk** [comment]
@johnmellor I'd also appreciate if you could file that IDL issue. The specification prose should operate on IDL primitives and ideally not call directly into JavaScript.

**Comment by johnmellor** [comment]
Thanks! I've addressed all comments till now, except that I left in `maxActions` since there doesn't seem to be consensus there yet.

I filed the IDL issue as https://www.w3.org/Bugs/Public/show_bug.cgi?id=29004

File: notifications.bs (hunk 4 above)
Comment (annevk) [inline notifications.bs]: I mostly care about a consistent story. Why is this different from vibrate?

Comment (johnmellor) [inline notifications.bs]: Historically, the DAP WG had fingerprinting concerns around exposing details of hardware vibrators.

By contrast, number of actions supported has much less fingerprinting risk, since it's only based on screen size and possibly OS, both of which are already exposed. Indeed Blink currently intends to always support 2 actions (both on Android and desktop).

Also, our thinking on fingerprinting has evolved, based on [data showing that it's infeasible to prevent fingerprinting](https://www.chromium.org/Home/chromium-security/security-faq#TOC-Why-isn-t-passive-browser-fingerprinting-including-passive-cookies-in-Chrome-s-threat-model-).

Comment (annevk) [inline notifications.bs]: But we are exposing the vibrate limit, just in a different way.

**Comment by johnmellor** [comment]
I've updated this PR, in particular:

- I realised that I can't share NotificationEvent with non-persistent
notifications, since it inherits from ExtendableEvent. Additionally, it's
unclear whether actions make sense on non-persistent notifications, which are
more toast-like. Hence actions can now only be provided for persistent
notifications. Passing actions for non-persistent notifications throws a
TypeError (though perhaps it would be more forward-compatible to do nothing?).

- I tried to explain the freezing of the actions attribute more precisely (based on
https://www.w3.org/Bugs/Public/show_bug.cgi?id=29004, it seems that using a
FrozenArray is still a reasonable choice, and freezing the values in the
FrozenArray doesn't seem totally crazy, though it remains unclear whether
there will be a WebIDL primitive for this).

- I rebased and ran Bikeshed.

- (I haven't changed maxActions, since I haven't heard any recent opinions for or
against).

[Updated HTML diff](http://services.w3.org/htmldiff?doc1=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2F90e29647469fb26b1f707cf93d8c787b92d2bbdc%2Fnotifications.html&doc2=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2F316a115f59a15e68150c624648d0453dfcd6deab%2Fnotifications.html)

Please take another look - thanks!

File: notifications.bs [hunk 8]
```diff
@@ -189,6 +209,10 @@ or vibration pattern that is not otherwise accessible to the end user.
   <li><p>If <var>options</var>'s <code>sticky</code> is true, set
   <var>notification</var>'s <a>sticky preference flag</a>.
 
+  <li><p>Set <var>notification</var>'s list of <a>actions</a> to the first
+  {{Notification/maxActions}} entries from <var>options</var>'s
```
Comment (annevk) [inline notifications.bs]: maxActions will have to be some concept if we want to use it like this. Referencing code directly doesn't work well since it can be overridden by script.

File: notifications.bs [hunk 9]
```diff
@@ -253,15 +277,23 @@ section of HTML. [[!HTML]]
...
 <p>User agents are expected to honor the Unicode semantics of the text of a
 <a lt="concept notification">notification</a>'s
-<a lt="concept title">title</a> and <a>body</a>. Each is expected to be
-treated as an independent set of one or more bidirectional algorithm paragraphs
-when displayed, as defined by the bidirectional algorithm's rules P1, P2, and
-P3, including, for instance, supporting the paragraph-breaking behavior of
-U+000A LINE FEED (LF) characters. For each paragraph of the
-<a lt="concept title">title</a> and <a>body</a>, the
+<a lt="concept title">title</a>, <a>body</a>, and the
+{{NotificationAction/title}} of each of its <a>actions</a>. Each is expected to
```
Comment (annevk) [inline notifications.bs]: Same problem here.

File: notifications.bs [hunk 10]
```diff
@@ -997,6 +1091,8 @@ urlPrefix: http://www.w3.org/TR/vibration/
     text: validate and normalize
   urlPrefix: #idl-def-; type: interface
     text: VibratePattern; url: VibratePattern
+urlPrefix: https://people.mozilla.org/~jorendorff/es6-draft.html#sec-object.; type: dfn
```
Comment (annevk) [inline notifications.bs]: We should probably use https://tc39.github.io/ecma262/ now. @domenic?

File: notifications.bs (hunk 8 above)
Comment (johnmellor) [inline notifications.bs]: Done.

File: notifications.bs (hunk 9 above)
Comment (johnmellor) [inline notifications.bs]: Done, though note that it made things much more verbose, for example rather than:
```html
  <li><p>Set <var>notification</var>'s list of <a>actions</a> to the first
  <var>n</var> entries from <var>options</var>'s <code>actions</code>, where
  <var>n</var> is the <a>maximum number of actions</a> supported.
```
I ended up with:
```html
  <li><p>Set <var>notification</var>'s list of <a>actions</a> to an empty list,
  then for each <var>entry</var> in <var>options</var>'s <code>actions</code>,
  up to the <a>maximum number of actions</a> supported (skip any excess
  entries), perform the following steps:

  <ol>
    <li><p>Let <var>action</var> be a new <a lt="actions">action</a>.

    <li><p>Set <var>action</var>'s <a>action name</a> to the <var>entry</var>'s
    <code>action</code>.

    <li><p>Set <var>action</var>'s <a lt="action title">title</a> to the
    <var>entry</var>'s <code>title</code>.

    <li><p>Append <var>action</var> to <var>notification</var>'s list of
    <a>actions</a>.
  </ol>
```
And similarly for the conversion back from concept to `NotificationAction`.

File: notifications.bs (hunk 10 above)
Comment (johnmellor) [inline notifications.bs]: Done.

**Comment by johnmellor** [comment]
[Updated HTML diff](http://services.w3.org/htmldiff?doc1=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2F90e29647469fb26b1f707cf93d8c787b92d2bbdc%2Fnotifications.html&doc2=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2Fd3aa9b127e8a47ba9fb136649b981874ed3cabf5%2Fnotifications.html)

**Comment by johnmellor** [comment]
Re-added explicit mention of creating a frozen array, to avoid confusion. [Updated HTML diff](http://services.w3.org/htmldiff?doc1=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2Ff72a0417b9cacc3e29194bc7e0e602d97224e5a5%2Fnotifications.html&doc2=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2Ffe9c3cce079fbdcf7d764c09ff829ee3f1d29f25%2Fnotifications.html).

File: notifications.bs [hunk 11]
```diff
@@ -100,6 +100,17 @@ associated <dfn>icon URL</dfn>, <dfn>icon resource</dfn>, <dfn>sound URL</dfn>,
 <p class=note>Developers are encouraged to not convey information through an icon, sound,
 or vibration pattern that is not otherwise accessible to the end user.
 
+<p>A <a lt="concept notification">notification</a> has an associated list of
+zero or more <dfn>actions</dfn>. Each action has an associated
+<dfn lt="action title">title</dfn> and <dfn>action name</dfn>. Users may
+activate actions, as alternatives to activating the notification itself. The
+user agent should determine the <dfn>maximum number of actions</dfn> supported,
```
Comment (annevk) [inline notifications.bs]: Must, no? Because if it doesn't, what would `maxActions` return?

File: notifications.bs [hunk 12]
```diff
@@ -343,20 +394,26 @@ language tag. Validity or well-formedness are not enforced. [[!LANG]]
...
 
 <ol>
   <li><p>If <var>notification</var> is a <a>persistent notification</a>, run
   these substeps:
 
   <ol>
+    <li><p>If one of <var>notification</var>'s <a>actions</a> was activated by
+    the user, let <var>action</var> be that action's <a>action name</a> string.
+
+    <li><p>Otherwise, let <var>action</var> be the empty string.
```
Comment (annevk) [inline notifications.bs]: I think I would prefer reversing these two steps and explicitly setting setting `<var>action</var>` if an action happened to be activated. Using "let" twice feels somewhat unclean (although I realize it's currently common in the HTML standard).

File: notifications.bs [hunk 13]
```diff
@@ -471,6 +530,7 @@ interface Notification : EventTarget {
   readonly attribute boolean requireInteraction;
   readonly attribute boolean sticky;
   [SameObject] readonly attribute any data;
+  readonly attribute FrozenArray&lt;NotificationAction> actions;
```
Comment (annevk) [inline notifications.bs]: This should be `[SameObject]`, no?

File: notifications.bs (hunk 7 above)
Comment (annevk) [inline notifications.bs]: Could you perhaps not modify this example as you are not changing it? Same for the next line. Omitting `;` is perfectly fine.

File: notifications.bs [hunk 14]
```diff
@@ -704,20 +773,75 @@ getter must return the <a lt="concept notification">notification</a>'s
...
+  <ol>
+    <li><p>Let <var>action</var> be a new {{NotificationAction}}.
+
+    <li><p>Set <var>action</var>'s {{NotificationAction/action}} to
+    <var>entry</var>'s <a>action name</a>.
+
+    <li><p>Set <var>action</var>'s {{NotificationAction/title}} to
+    <var>entry</var>'s <a lt="action title">title</a>.
+
+    <li><p>Call <a lt="freeze">Object.freeze</a> on <var>action</var>.
```
Comment (annevk) [inline notifications.bs]: I think you need to say here that you are using the initial value of `Object.freeze` so it cannot possibly be overridden from script. There should probably also be some XXX type of comment in the source here and a bug on IDL since it's somewhat strange to invoke JavaScript on what are effectively IDL constructs.

File: notifications.bs (hunk 11 above)
Comment (johnmellor) [inline notifications.bs]: Done.

File: notifications.bs (hunk 12 above)
Comment (johnmellor) [inline notifications.bs]: Done.

File: notifications.bs (hunk 13 above)
Comment (johnmellor) [inline notifications.bs]: Done (indeed, there's currently no reason the UA would replace actions with a different FrozenArray).

File: notifications.bs (hunk 14 above)
Comment (johnmellor) [inline notifications.bs]: Done (I've added an XXX comment, which links to https://www.w3.org/Bugs/Public/show_bug.cgi?id=29004, and explained in the prose that this is "to prevent accidental mutation by scripts").

**Comment by johnmellor** [comment]
[Updated HTML diff](http://services.w3.org/htmldiff?doc1=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2Ff72a0417b9cacc3e29194bc7e0e602d97224e5a5%2Fnotifications.html&doc2=https%3A%2F%2Frawgit.com%2Fjohnmellor%2Fnotifications%2F5022b8399443601162f77557d023e80411a15154%2Fnotifications.html)

File: notifications.bs (hunk 7 above)
Comment (johnmellor) [inline notifications.bs]: The example used inconsistent variable names, `var not` then `notification.onclick`. That seems worth fixing, though I'm happy to do that as a separate PR if you'd prefer? Left it in for now.

File: notifications.bs [hunk 15]
```diff
@@ -189,6 +212,24 @@ or vibration pattern that is not otherwise accessible to the end user.
   <li><p>If <var>options</var>'s <code>sticky</code> is true, set
   <var>notification</var>'s <a>sticky preference flag</a>.
 
+  <li><p>Set <var>notification</var>'s list of <a>actions</a> to an empty list,
```
Comment (annevk) [inline notifications.bs]: Sorry for not spotting this earlier. Since the `<p>` is not the only child of the `<li>` it needs to appear on its own line and indented. Same for the `<ol>` (and its children).

File: notifications.bs [hunk 16]
```diff
@@ -704,20 +773,81 @@ getter must return the <a lt="concept notification">notification</a>'s
...
 <a lt="concept notification">notification</a>'s <a>data</a>.
 
+<p>The <dfn attribute dfn-for=Notification><code>actions</code></dfn>
+attribute's getter must return the result of the following steps:
+
+<ol>
+  <li><p>Let <var>frozenActions</var> be an empty list of type
+  {{NotificationAction}}.
+
+  <li><p>For each <var>entry</var> in the
```
Comment (annevk) [inline notifications.bs]: Same issue with multiple children here.

File: notifications.bs (hunk 7 above)
Comment (annevk) [inline notifications.bs]: I see. That seems worth fixing. I guess I can live with the added semicolons.

**Comment by annevk** [comment]
@sicking, does this look good to you too?

File: notifications.bs (hunk 15 above)
Comment (johnmellor) [inline notifications.bs]: Hmm, the prevailing style seems to be not to indent these - see [line 289][1],  [line 306][2],  [line 351][3],  [line 369][4], and  [line 413][5].

[1]: https://github.com/whatwg/notifications/blob/f72a0417b9cacc3e29194bc7e0e602d97224e5a5/notifications.bs#L289
[2]: https://github.com/whatwg/notifications/blob/f72a0417b9cacc3e29194bc7e0e602d97224e5a5/notifications.bs#L306
[3]: https://github.com/whatwg/notifications/blob/f72a0417b9cacc3e29194bc7e0e602d97224e5a5/notifications.bs#L351
[4]: https://github.com/whatwg/notifications/blob/f72a0417b9cacc3e29194bc7e0e602d97224e5a5/notifications.bs#L369
[5]: https://github.com/whatwg/notifications/blob/f72a0417b9cacc3e29194bc7e0e602d97224e5a5/notifications.bs#L413

The only examples I can find of indenting the children - [line 434][6],  [line 581][7], and  [line 594][8] - all use single-space indent for the entire enclosing `<ol>`. But I'd rather not convert other `<ol>`s like the giant [create-a-notification](https://notifications.spec.whatwg.org/#create-a-notification) `<ol>` to single-space indent.

[6]: https://github.com/whatwg/notifications/blob/f72a0417b9cacc3e29194bc7e0e602d97224e5a5/notifications.bs#L434
[7]: https://github.com/whatwg/notifications/blob/f72a0417b9cacc3e29194bc7e0e602d97224e5a5/notifications.bs#L581
[8]: https://github.com/whatwg/notifications/blob/f72a0417b9cacc3e29194bc7e0e602d97224e5a5/notifications.bs#L594

Do still you want me to indent these? And if so, shall I use 2-space indent even though there's no precedent for that?

Comment (annevk) [inline notifications.bs]: Hmm, I guess it's fine to leave this for now then. Thank you for the research. Should probably be cleaned up at some point, but given the precedent it's not really important to do it as part of this.

**Comment by johnmellor** [comment]
(I reordered the `<li>`s I added to the `create a notification` algorithm, since it's [easier to implement](https://codereview.chromium.org/1409143004/diff/1/third_party/WebKit/Source/modules/notifications/Notification.cpp#newcode82) when the steps that only apply to non-persistent notifications aren't interleaved with the parts of the algorithm that apply to both).

**Comment by sicking** [comment]
I looked through the HTML-diff, mainly looking at the WebIDL changes, and this looks great to me!
